consensus/bundle_validator/bundle_validator.c

```diff
@@ -0,0 +1,220 @@
+    return res;
+  }
+
+  size_t last_index = curr_tx->last_index;
+  size_t curr_index = 0;
+  memcpy(bundle_hash, curr_tx->bundle, FLEX_TRIT_SIZE_243);
+  trit_array_t curr_tx_trunk;
+  curr_tx_trunk.dynamic = 0;
+  curr_tx_trunk.trits = curr_tx_trunk_trits;
+  while (curr_index < last_index &&
+         memcmp(bundle_hash, curr_tx->bundle, FLEX_TRIT_SIZE_243) == 0) {
+    bundle_transactions_add(bundle, curr_tx);
+    pack.num_loaded = 0;
+    pack.insufficient_capacity = false;
+    trit_array_set_trits(&curr_tx_trunk, curr_tx->trunk, NUM_TRITS_HASH);
+    res = iota_tangle_transaction_load(tangle, TRANSACTION_COL_HASH,
+                                       &curr_tx_trunk, &pack);
+
+    if (res != RC_OK || pack.num_loaded == 0) {
+      log_error(BUNDLE_VALIDATOR_ID,
+                "No transactions were loaded for the provided tail hash\n");
+      return res;
+    }
+  }
+
+  return RC_OK;
+}
+
+retcode_t validate_signature(bundle_transactions_t* bundle,
+                             trit_t* normalized_bundle, bool* is_valid) {
+  retcode_t res = RC_OK;
+  iota_transaction_t curr_tx;
+  iota_transaction_t curr_inp_tx;
+  Kerl address_kerl;
+  Kerl sig_frag_kerl;
+  trit_t digested_sig_trits[NUM_TRITS_ADDRESS];
+  trit_t digested_address[NUM_TRITS_ADDRESS];
+  trit_t tx_address_trits[NUM_TRITS_ADDRESS];
+  trit_t key[NUM_TRITS_SIGNATURE];
+  size_t i = 0;
+
+  init_kerl(&address_kerl);
+  init_kerl(&sig_frag_kerl);
+
+  for (curr_tx = (iota_transaction_t)utarray_eltptr(bundle, 0);
+       curr_tx != NULL;) {
+    if (curr_tx->value >= 0) {
+      curr_tx = (iota_transaction_t)utarray_next(bundle, curr_tx);
+      continue;
+    }
+    curr_inp_tx = curr_tx;
+    size_t offset = 0, next_offset = 0;
+    kerl_reset(&address_kerl);
+    do {
+      kerl_reset(&sig_frag_kerl);
+      next_offset = (offset + ISS_FRAGMENTS * RADIX - 1) % NUM_TRITS_HASH + 1;
+      flex_trits_to_trits(key, NUM_TRITS_SIGNATURE,
+                          curr_inp_tx->signature_or_message,
+                          NUM_TRITS_SIGNATURE, NUM_TRITS_SIGNATURE);
+
+      iss_kerl_sig_digest(digested_sig_trits,
+                          &normalized_bundle[offset % NUM_TRITS_HASH], key,
+                          NUM_TRITS_SIGNATURE, &sig_frag_kerl);
+      kerl_absorb(&address_kerl, digested_sig_trits, NUM_TRITS_ADDRESS);
+      curr_inp_tx = (iota_transaction_t)utarray_next(bundle, curr_inp_tx);
+      offset = next_offset;
+    } while (curr_inp_tx != NULL, memcmp(curr_inp_tx->address, curr_tx->address,
+                                         FLEX_TRIT_SIZE_243) == 0 &&
+                                      curr_inp_tx->value == 0);
+    kerl_squeeze(&address_kerl, digested_address, NUM_TRITS_ADDRESS);
+
+    flex_trit_t digest[FLEX_TRIT_SIZE_243];
+
+    flex_trits_from_trits(digest, NUM_TRITS_HASH, digested_address,
+                          NUM_TRITS_HASH, NUM_TRITS_HASH);
+    flex_trits_to_trits(tx_address_trits, NUM_TRITS_ADDRESS, curr_tx->address,
```

Comment:
Why are we overwriting the address on the current tx here?

Reply:
we are converting the address to trits, not overriding it, but still, there is one redundant conversion here that got away..